dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/config/HibernateConfig.java

```diff
@@ -67,62 +81,121 @@ public HibernateConfigurationProvider hibernateConfigurationProvider(
 
+  @Bean("sessionFactory")
+  @Primary
+  public SessionFactory sessionFactory(
+      @Qualifier("entityManagerFactory") EntityManagerFactory entityManager) {
+    return entityManager.unwrap(SessionFactory.class);
+  }
+
   @Bean
   public DbmsManager dbmsManager(
       JdbcTemplate jdbcTemplate,
-      SessionFactory sessionFactory,
-      DefaultHibernateCacheManager cacheManager) {
+      DefaultHibernateCacheManager cacheManager,
+      EntityManager entityManager) {
     HibernateDbmsManager hibernateDbmsManager = new HibernateDbmsManager();
     hibernateDbmsManager.setCacheManager(cacheManager);
-    hibernateDbmsManager.setSessionFactory(sessionFactory);
+    hibernateDbmsManager.setEntityManager(entityManager);
     hibernateDbmsManager.setJdbcTemplate(jdbcTemplate);
     return hibernateDbmsManager;
   }
+
+  @Bean
+  public BeanFactoryPostProcessor entityManagerBeanDefinitionRegistrarPostProcessor() {
+    return new EntityManagerBeanDefinitionRegistrarPostProcessor();
+  }
+
+  @Bean("entityManagerFactory")
+  @DependsOn({"flyway"})
+  public EntityManagerFactory entityManagerFactoryBean(
+      DhisConfigurationProvider dhisConfig, DataSource dataSource) {
+    HibernateJpaVendorAdapter adapter = new HibernateJpaVendorAdapter();
+    adapter.setDatabasePlatform(dhisConfig.getProperty(ConfigurationKey.CONNECTION_DIALECT));
+    adapter.setGenerateDdl(shouldGenerateDDL(dhisConfig));
+    LocalContainerEntityManagerFactoryBean factory = new LocalContainerEntityManagerFactoryBean();
+    factory.setJpaVendorAdapter(adapter);
+    factory.setPersistenceUnitName("dhis");
+    factory.setPersistenceProvider(new org.hibernate.jpa.HibernatePersistenceProvider());
+    factory.setDataSource(dataSource);
+    factory.setPackagesToScan("org.hisp.dhis");
+    factory.setSharedCacheMode(SharedCacheMode.ENABLE_SELECTIVE);
+    factory.setValidationMode(ValidationMode.AUTO);
+    factory.setJpaProperties(getAdditionalProperties());
+    factory.setMappingResources(loadResources());
+    factory.afterPropertiesSet();
+    return factory.getObject();
+  }
+
+  /**
+   * Returns additional properties to be used by the {@link LocalContainerEntityManagerFactoryBean}
+   */
+  private Properties getAdditionalProperties() {
+    Properties properties = new Properties();
+    properties.put(
+        "hibernate.current_session_context_class",
+        "org.springframework.orm.hibernate5.SpringSessionContext");
+
+    // TODO: this is anti-pattern and should be turn off
+    properties.put("hibernate.allow_update_outside_transaction", "true");
+
+    return properties;
+  }
+
+  /**
+   * Loads all the hibernate mapping files from the classpath
+   *
+   * @return Array of Strings representing the mapping files
+   */
+  private String[] loadResources() {
+    try {
+      PathMatchingResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();
+      Resource[] resources = resolver.getResources("classpath*:org/hisp/dhis/**/*.hbm.xml");
```

Comment:
Could this file check be more targeted? It looks like it's going to look through every package on the classpath when we know exactly where all `hbm.xml` files are located.

Reply:
Not sure how to do this. Will check later and update on next PR.